keps/sig-node/20190425-node-readiness-gates.md

```diff
@@ -0,0 +1,388 @@
+
+As Kubernetes becomes more and more customizable (or extensible) and gets deployed in mission critical systems,
+it is imperative to provide fail safe mechanisms that are built into the system.
+
+The remainder of this proposal introduces an approach for addressing the issue of node health extensibility using
+node readiness gates.
+
+### Goals
+
+* introduce a mechanism that allows for customizable/extensible node readiness checks.
+
+### Non-Goals
+
+* improving time-to-readiness of nodes
+
+## Proposal
+
+Any solution that is chosen for addressing the set of problems stated above should meet the following design requirements:
+
+* Be backwards compatible
+* Be extensible to accommodate variations in kubernetes environments - the set of system microservices can vary across k8s deployments.
+* Ideally, provide a single signal (not a policy) to cluster level components.
+* Ideally, avoid introducing new API primitives.
+* Support self-bootstrapping - scheduling system components (including the kubelet) even if the node is not logically ready.
+
+## Design Details
+
+### Determining Readiness via Pod Selectors
+
+In this design, the indication of node readiness can be expressed by the readiness/health of specific pods running on that node. This approach
+would leverage existing mechnanisms and well-known concepts to determine if a node is ready to receive pods.
+
+A readiness taint will be used to track the overall readiness of a node. The kubelet, node lifecycle controller, other external controllers will
+manage that taint.
+
+The Node object will be extended to include a list of readiness gates like so:
+
+```go
+type NodeSpec struct {
+  ...
+  // Defines the list of readiness gates for a node
+  ReadinessGates []NodeReadinessGate `json:"readinessGates,omitempty" protobuf:"bytes,7,opt,name=readinessGates"`
+}
+
+// NodeReadinessGate indicates a set of pods that must be ready on a given node before the
+// node itself is considered ready. Pods listed under NodeReadinessGates should be system critical pods
+// that should block other workloads from running on a node until it is ready.
+type NodeReadinessGate struct {
+   // The name of the readiness gate, mainly required as the patch merge key
+   Name string `json:"name,omitempty" protobuf:"bytes,1,opt,name=name"`
+   // A label query for pods representing this readiness gate.
+   // Generally, labels here should point to a specific pod on a node.
+   Selector LabelSelector `json:"selector,omitempty" protobuf:"bytes,2,opt,name=selector"`
+   // The namespace query for the pods representing this readiness gate.
+   Namespace string `json:"namespace,omitempty" protobuf:"bytes,3,opt,name=namespace"`
+   // Indicates if this readiness gate should block the overall readiness of a node. Default: true
+   // +optional
+   BlocksReadiness bool `json:"blocksReadiness,omitempty" protobuf:"varint,4,opt,name=blocksReadiness"`
```

Comment:
I see a use-case for this, but not 100% sold on it. Thoughts?

Reply:
I agree, I'd like to see clear use-cases for this to add it